### PR TITLE
Add complex types to Metadata

### LINKED ISSUE #13947: Implement complex types (value objects)
Greetings, this is a question/feature request:

Per the Limitations section of the Value Conversions documentation: "There is currently no way to spread a conversion of one property to multiple columns or vice-versa."

Is something like this planned, or is there an alternative better way to do this?
I'm looking for a way to keep the backing fields out of my entity classes. And to not use strings in the configuration or public backing fields in my entity classes.

Here's a quick example from my current code working with a legacy DB:

From an entity class:

```c#
      private DateTime _lastChanged { get; set; }
      private int _lastChangedHour { get; set; }
      private int _lastChangedMinute { get; set; }      
      public DateTime LastChanged
      {
         get
         {
            if (_lastChanged.Hour != 0 || 
                _lastChanged.Minute != 0 || 
                _lastChanged.Second != 0)
               return _lastChanged;

            var returnDateTime = 
               new DateTime(
                  _lastChanged.Year, 
                  _lastChanged.Month, 
                  _lastChanged.Day, 
                  _lastChangedHour, 
                  _lastChangedMinute, 
                  0);
            return returnDateTime;
         }

         set
         {
            _lastChangedHour = value.Hour;
            _lastChangedMinute = value.Minute;
            _lastChanged = value;
         }
      }
```

From the entity's configuration:

```c#
            builder.Ignore(o => o.LastChanged);

            builder
               .Property<DateTime>("_lastChanged")
               .HasColumnName("SIST_ENDRET")
               .HasColumnType("datetime")
               .HasDefaultValue((DateTime) SqlDateTime.MinValue);

            builder
               .Property<int>("_lastChangedHour")
               .HasColumnName("SIST_ENDRET_TIM")
               .HasColumnType("decimal(2,0)")
               .HasConversion<decimal>();

            builder
               .Property<int>("_lastChangedMinute")
               .HasColumnName("SIST_ENDRET_MIN")
               .HasColumnType("decimal(2,0)")
               .HasConversion<decimal>();
```

Thanks for your time!

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Fixes #13947

### CODE REVIEW

**Comment by AndriySvyryd** [comment]
I'll be adding more tests, but this is ready to be reviewed.

File: src/EFCore/Storage/ITypeMappingSource.cs [hunk 1]
```diff
@@ -30,19 +30,19 @@ namespace Microsoft.EntityFrameworkCore.Storage;
 public interface ITypeMappingSource
 {
     /// <summary>
-    ///     Finds the type mapping for a given <see cref="IProperty" />.
+    ///     Finds the type mapping for a given <see cref="IPrimitivePropertyBase" />.
     /// </summary>
     /// <param name="property">The property.</param>
     /// <returns>The type mapping, or <see langword="null" /> if none was found.</returns>
-    CoreTypeMapping? FindMapping(IProperty property);
+    CoreTypeMapping? FindMapping(IPrimitivePropertyBase property);
```
Comment (AndriySvyryd) [inline src/EFCore/Storage/ITypeMappingSource.cs:37]: @ajcvickers Breaking change

File: src/EFCore/Metadata/Builders/IConventionPrimitivePropertyBaseBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+///     </para>
+///     <para>
+///         This interface is typically used by database providers (and other extensions). It is generally
+///         not used in application code.
+///     </para>
+/// </summary>
+/// <remarks>
+///     See <see href="https://aka.ms/efcore-docs-conventions">Model building conventions</see> for more information and examples.
+/// </remarks>
+public interface IConventionPrimitivePropertyBaseBuilder<TBuilder> : IConventionPropertyBaseBuilder<TBuilder>
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/IConventionPrimitivePropertyBaseBuilder.cs:20]: Does a generic base make sense?

File: src/EFCore/Metadata/IComplexProperty.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace Microsoft.EntityFrameworkCore.Metadata;
+
+/// <summary>
+///     Represents a complex property of a structural type.
+/// </summary>
+/// <remarks>
+///     See <see href="https://aka.ms/efcore-docs-modeling">Modeling entity types and relationships</see> for more information and examples.
+/// </remarks>
+public interface IComplexProperty : IReadOnlyComplexProperty, IPropertyBase
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/IComplexProperty.cs:12]: Should there be a `IComplexTypeComplexProperty`?

File: src/EFCore/Storage/ITypeMappingSource.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore/Storage/ITypeMappingSource.cs:37]: Triage decided to merge `IPrimitivePropertyBase` and `IComplexcTypeProperty` to `IProperty`

File: src/EFCore/Metadata/Builders/IConventionPrimitivePropertyBaseBuilder.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Builders/IConventionPrimitivePropertyBaseBuilder.cs:20]: Triage decided to merge `IPrimitivePropertyBase` and `IComplexTypeProperty` to `IProperty`

File: src/EFCore/Metadata/IComplexProperty.cs (hunk 3 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/IComplexProperty.cs:12]: Triage decided to merge `IPrimitivePropertyBase` and `IComplexTypeProperty` to `IProperty`

File: src/EFCore.InMemory/ValueGeneration/Internal/InMemoryValueGeneratorSelector.cs [hunk 4]
```diff
@@ -58,82 +58,66 @@ private ValueGenerator GetOrCreate(IProperty property)
...
 
     private bool FindGenerator(IProperty property, Type type, out ValueGenerator? valueGenerator)
     {
         if (type == typeof(long))
         {
-            {
-                valueGenerator = _inMemoryStore.GetIntegerValueGenerator<long>(property);
-                return true;
-            }
+            valueGenerator = _inMemoryStore.GetIntegerValueGenerator<long>(property);
```
Comment (ajcvickers) [inline src/EFCore.InMemory/ValueGeneration/Internal/InMemoryValueGeneratorSelector.cs:68]: How did this formatting even happen? :-)

**Review by ajcvickers — APPROVED** [review]
🐑 🇮🇹

**Pull request merged**